lib/reservation_manager.rb

```diff
@@ -0,0 +1,204 @@
+        reservation = make_reservation(checkin_date, checkout_date, 1, @room_block_discount)
+        reservations << reservation
+        room_numbers << reservation.room_number
+      end
+
+      block_data = parse_room_block_data(checkin_date, checkout_date, num_rooms, room_numbers)
+
+      room_block =  Hotel::RoomBlock.new(block_data)
+      add_room_block_to_list(room_block)
+
+      return room_block
+    end
+
+    #returns total cost for a given reservation
+    def get_total_reservation_cost(given_confirmation_id)
+
+      return @reservations.reduce(0) do |cost, reservation|
+        if reservation.confirmation_id == given_confirmation_id.to_s
+          cost + reservation.total_cost
+        end
+      end
+    end
+
+    #returns array of list of room numbers in all rooms
+    def list_all_rooms_in_hotel
+
+      return @rooms.map { |room| room.room_number }
+    end
+
+    #helper method for listing all available rooms
+    def find_all_available_rooms(given_date_range)
+
+      available_rooms = []
+
+      @rooms.find_all do |room|
+        if (room.is_available?(given_date_range))
+          available_rooms << room.room_number
+        end
+      end
+
+      return "Hotel fully booked for this date range. Try a different date." if available_rooms.empty?
+      return available_rooms
+    end
+
+    #returns array of rooms available today
+    def list_rooms_available_today
+
+      return find_all_available_rooms([Date.today()])
+    end
+
+    #returns array of rooms available for a given date
+    def list_available_rooms(given_date_range)
+
+      return find_all_available_rooms(given_date_range)
+    end
+
+    # returns an array of the list of reservations for a specific date
+    # won't return the reservation if checkout on is on the given_day
+    def list_daily_reservations(given_date)
+
+      given_date = Date.parse(given_date)
+      daily_reservations = []
+
+      @reservations.find_all do |reservation|
+        if ((reservation.checkin_date..reservation.checkout_date).to_a - [reservation.checkout_date]).include?(given_date)
+          daily_reservations << reservation.confirmation_id
+        end
+      end
+
+      return "No reservations for #{given_date}." if daily_reservations.empty?
+      return daily_reservations
+    end
+
+    private
+
+    #private method to generate valid room ids for ReservationManager
+    def generate_room_ids(room_ids)
+      @rooms = room_ids.map do |room_id|
+        Room.new(room_id)
+      end
+      raise StandardError.new('No rooms remaining to create.') if room_ids.empty?
+    end
+
+    #private method to generate reservation ids according to naming convention constant
+    def generate_random_reservation_id
+
+      return (('A'..'Z').to_a.sample(3) + (0..9).to_a.sample(5)).join
+    end
+
+    #private method to generate block ids according to naming convention constant
+    def generate_random_block_id
+      return ((0..9).to_a.sample(5)).join
```

Comment:
Just an observation: this approach with `sample` means we'll never generate a block ID with multiple of the same digit. This goes for the reservation ID generation code as well.